### DISCUSSION TITLE
Renovate cannot autodiscover repositories (Bitbucket Server / On Prem)

### DISCUSSION THREAD

**Discussion by uhthomas** [renovatebot/renovate#17776] [Request Help]
The result of `getRepos` is always empty, which is really strange.

```json
{"name":"renovate","hostname":"renovate-whitesource-renovate-7946d4dc89-xqc65","pid":2,"level":20,"logContext":"z0tJfPxBh2dtiNfXrBfAX","result":[],"msg":"result of getRepos()","time":"2022-09-13T22:05:00.020Z","v":0}
```

The username and token are fine as a curl works as expected:

```sh
curl -u 'renovate-bot:http_access_token' 'https://bitbucket.example.com/rest/api/1.0/repos?permission=REPO_WRITE&state=AVAILABLE'
```

```json
{
        "size": 25,
        "limit": 25,
        "isLastPage": false,
        "values": [
                { "slug": "..." },
         ],
        "start": 0,
        "nextPageStart": 25
}
```

Could anyone help understand what's going wrong?

Using https://hub.docker.com/layers/whitesource/renovate/2.5.1/images/sha256-5e14ab06a9107dd428d2ecb32ef5a29cfea4fb6b4cba7b6a2c91cc142255ce8e?context=explore

**Comment by viceice** [comment]
Do you configure the token as `password` or as `token`?

https://docs.renovatebot.com/modules/platform/bitbucket-server/

**Reply by uhthomas** [reply]
I'm using `RENOVATE_USERNAME` and `RENOVATE_PASSWORD`. I would have assumed an error would be raised if either of these values were to be incorrect?

**Reply by viceice** [reply]
Then there isn't much we can do. Maybe some server side restrictions / rate limiting?

https://github.com/renovatebot/renovate/blob/64a58171dff2023586cb5583d9f90ea0570a1c01/lib/modules/platform/bitbucket-server/index.ts#L105-L121

**Reply by uhthomas** [reply]
I don't understand why a `curl` to the same endpoint with the same credentials works as expected. Is Renovate not catching an error somewhere, or is it failing to parse the response for some reason? Is there anything we can do to debug further? We're using Bitbucket `v7.21.4`.

**Reply by uhthomas** [reply]
FWIW I tried changing the username and password to something invalid and got an expected authentication failure. This tells us that Renovate can communicate with Bitbucket and authenticate. So, what's going wrong? I've also tried `2.4.1` and `2.5.0` with the same result...

**Reply by uhthomas** [reply]
Checking our logs I can see logs for `GET /rest/api/1.0/users/renovate-bot HTTP/1.1` and some logs for `"GET /rest/api/1.0/repos HTTP/1.1"` where I ran curl manually. There are less than 30 matches for `renovate-bot` which does not match the amount of logs for `getRepos` from Renovate for which there are over 1000 in the past 24 hours.

Given this, it sounds like the request isn't even hitting Bitbucket, maybe? Is there something bad happening when the URL is transformed in `accumulateValues`?

**Reply by uhthomas** [reply]
Just for clarification, we know that Renovate _can_ communicate with Bitbucket as explained before, though for some reason it seems like it's not even making a request with `getRepos`.

**Reply by viceice** [reply]
it seems someone needs to debug this, i no longer have a bitbucket server, so can't do it myself.

**Reply by uhthomas** [reply]
Ah, I see. Thanks @viceice. Who can help debug this?

**Reply by rarkins** [reply]
What exact software are you running?

**Reply by uhthomas** [reply]
Using https://github.com/whitesource/renovate-on-prem in Kubernetes with [whitesource/renovate:2.5.1](https://hub.docker.com/layers/whitesource/renovate/2.5.1/images/sha256-5e14ab06a9107dd428d2ecb32ef5a29cfea4fb6b4cba7b6a2c91cc142255ce8e?context=explore) and Bitbucket Server `v7.21.4`.

**Reply by rarkins** [reply]
Sorry I didn't notice it earlier but it doesn't support Bitbucket - only GitHub and GitLab

**Reply by uhthomas** [reply]
Ah, I see. Thanks @rarkins. I suppose the solution is to use https://github.com/renovatebot/helm-charts/tree/main/charts/renovate instead?

What are the differences? What's blocking https://github.com/whitesource/renovate-on-prem/issues/6?

**Reply by rarkins** [reply]
> Ah, I see. Thanks @rarkins. I suppose the solution is to use https://github.com/renovatebot/helm-charts/tree/main/charts/renovate instead?

Yes, use the open source Renovate instead

> What are the differences? What's blocking [whitesource/renovate-on-prem#6](https://github.com/whitesource/renovate-on-prem/issues/6)?

The main blocker (apart from time) is webhooks. Bitbucket Server doesn't have an elegant solution for us to listen on them like we can for github and gitlab

**Reply by uhthomas** [reply]
@rarkins Is the "Web Post Hooks" plugin a viable solution to support Bitbucket Server?

https://marketplace.atlassian.com/apps/1211539/web-post-hooks-for-bitbucket-server?hosting=server&tab=overview

**Reply by rarkins** [reply]
![image](https://user-images.githubusercontent.com/6311784/199489651-a15b128d-453a-4356-95f0-fbbf68a0f559.png)

**Reply by uhthomas** [reply]
Ah, thanks @rarkins! Sounds like Bitbucket Server should support webhooks natively then? Does this unblock things for Renovate?

**Reply by rarkins** [reply]
Potentially, but because BB server has been deprecated it lowers its priority compared to other platforms, so we don't have it on our roadmap
